flang/lib/Optimizer/Transforms/DoConcurrentConversion.cpp

```diff
@@ -99,91 +75,379 @@ class DoConcurrentConversion : public mlir::OpConversionPattern<fir::DoLoopOp> {
+
+    mlir::Value shape = [&]() -> mlir::Value {
+      if (hlfir::isFortranScalarNumericalType(liveInType))
+        return {};
+
+      if (hlfir::isFortranArrayObject(liveInType)) {
+        llvm::SmallVector<mlir::Value> shapeOpOperands;
 
-    // Collect the memref defining ops in the parallel region.
-    for (mlir::Operation *opToClone : declareAndAllocasToClone) {
+        for (auto boundsOperand : liveInMapInfoOp.getBounds()) {
+          auto boundsOp =
+              mlir::cast<mlir::omp::MapBoundsOp>(boundsOperand.getDefiningOp());
+          mlir::Operation *localExtentDef =
+              boundsOp.getExtent().getDefiningOp()->clone();
+          rewriter.getInsertionBlock()->push_back(localExtentDef);
+          assert(localExtentDef->getNumResults() == 1);
+
+          shapeOpOperands.push_back(localExtentDef->getResult(0));
+        }
+
+        return rewriter.create<fir::ShapeOp>(liveInArg.getLoc(),
+                                             shapeOpOperands);
+      }
+
+      std::string opStr;
+      llvm::raw_string_ostream opOs(opStr);
+      opOs << "Unsupported type: " << liveInType;
+      llvm_unreachable(opOs.str().c_str());
+    }();
+
+    return rewriter.create<hlfir::DeclareOp>(liveInArg.getLoc(), liveInArg,
+                                             liveInMapInfoOp.getName().value(),
+                                             shape);
+  }
+
+  mlir::omp::TeamsOp
+  genTeamsOp(mlir::Location loc, mlir::ConversionPatternRewriter &rewriter,
+             fir::DoLoopOp doLoop, llvm::ArrayRef<mlir::Value> liveIns,
+             mlir::IRMapping &mapper,
+             mlir::omp::LoopNestClauseOps &loopNestClauseOps) const {
+    auto teamsOp = rewriter.create<mlir::omp::TeamsOp>(
+        loc, /*clauses=*/mlir::omp::TeamsClauseOps{});
+
+    mlir::Block *teamsBlock = rewriter.createBlock(&teamsOp.getRegion());
+    rewriter.create<mlir::omp::TerminatorOp>(loc);
+    rewriter.setInsertionPointToStart(teamsBlock);
+
+    genInductionVariableAlloc(rewriter, liveIns, mapper);
+    genLoopNestClauseOps(loc, rewriter, doLoop, mapper, loopNestClauseOps);
+
+    return teamsOp;
+  }
+
+  void
+  genLoopNestClauseOps(mlir::Location loc,
+                       mlir::ConversionPatternRewriter &rewriter,
+                       fir::DoLoopOp doLoop, mlir::IRMapping &mapper,
+                       mlir::omp::LoopNestClauseOps &loopNestClauseOps) const {
+    assert(loopNestClauseOps.loopLBVar.empty() &&
+           "Loop nest bounds were already emitted!");
+
+    // Clones the chain of ops defining a certain loop bound or its step into
+    // the parallel region. For example, if the value of a bound is defined by a
+    // `fir.convert`op, this lambda clones the `fir.convert` as well as the
+    // value it converts from. We do this since `omp.target` regions are
+    // isolated from above.
+    std::function<mlir::Operation *(mlir::Operation *)>
+        cloneBoundOrStepDefChain = [&](mlir::Operation *operation) {
```

Comment:
```suggestion
    auto cloneBoundOrStepDefChain = [&](mlir::Operation *operation) -> mlir::Operation * {
```

Reply:
It is a recursive lambda, you have to specify the type.